content/docs/intro/concepts/resources.md

```diff
@@ -0,0 +1,2654 @@
+
+```go
+ctx.RegisterResourceOutputs(myComponent, pulumi.Map{
+    "bucketDnsName": bucket.BucketDomainName,
+})
+```
+
+{{% /choosable %}}
+{{% choosable language csharp %}}
+
+```csharp
+this.RegisterOutputs(new Dictionary<string, object>
+{
+    { "bucketDnsName", bucket.BucketDomainName }
+});
+```
+
+{{% /choosable %}}
+
+{{< /chooser >}}
+
+The call to `registerOutputs` typically happens at the very end of the component resource’s constructor.
+
+The call to `registerOutputs` also tells Pulumi that the resource is done registering children and should be considered fully constructed, so—although it’s not enforced—the best practice is to call it in all components even if no outputs need to be registered.
+
+### Inheriting Resource Providers
+
+One option all resources have is the ability to pass an [explicit resource provider]({{< relref "#providers">}}) to supply explicit configuration settings. For instance, you may want to ensure that all AWS resources are created in a different region than the globally configured region. In the case of component resources, the challenge is that these providers must flow from parent to children.
+
+To allow this, component resources accept a `providers` option that custom resources don’t have. This value contains a map from the provider name to the explicit provider instance to use for the component resource. The map is used by a component resource to fetch the proper `provider` object to use for any child resources. This example overrides the globally configured AWS region and sets it to us-east-1. Note that `myk8s` is the name of the Kubernetes provider.
+
+{{< chooser language "javascript,typescript,python,go,csharp" >}}
+
+{{% choosable language javascript %}}
+
+```javascript
+let component = new MyComponent("...", {
+    providers: {
+        aws: useast1,
+        kubernetes: myk8s,
+    },
+});
+```
+
+{{% /choosable %}}
+{{% choosable language typescript %}}
+
+```typescript
+let component = new MyComponent("...", {
+    providers: {
+        aws: useast1,
+        kubernetes: myk8s,
+    },
+});
+```
+
+{{% /choosable %}}
+{{% choosable language python %}}
+
+```python
+component = MyComponent('...', ResourceOptions(providers={
+    'aws': useast1,
+    'kubernetes': myk8s,
+}))
+```
+
+{{% /choosable %}}
+{{% choosable language go %}}
+
+```go
+component, err := NewMyResource(ctx, "...", nil, pulumi.ProviderMap(
+    map[string]pulumi.ProviderResource{
+        "aws":        awsUsEast1,
+        "kubernetes": myk8s,
+    },
+))
+```
+
+{{% /choosable %}}
+{{% choosable language csharp %}}
+
+```csharp
+var component = new MyResource("...", new ComponentResourceOptions {
+    Providers = {
+        { "aws", awsUsEast1 },
+        { "kubernetes", myk8s }
+    }
+});
+```
+
+{{% /choosable %}}
+
+{{< /chooser >}}
+
+If a component resource is itself a child of another component resource, its set of providers is inherited from its parent by default.
+
+## Resource Providers {#providers}
```

Comment:
Wouldn't be surprised if we end up wanting to pull this out into it's own top-level section on "Providers" - especially as we flesh out more on how to build your own providers (beyond just Dynamic Providers as currently documented here).  But seems fine to leave in here for now.